numpyro/infer/elbo.py

```diff
@@ -577,34 +543,47 @@ def single_particle_elbo(rng_key):
             check_model_guide_match(model_trace, guide_trace)
             _validate_model(model_trace, plate_warning="strict")
 
-            # XXX: different from Pyro, we don't support baseline_loss here
-            non_reparam_nodes = {
-                name
-                for name, site in guide_trace.items()
-                if site["type"] == "sample"
-                and (not site["is_observed"])
-                and (not site["fn"].has_rsample)
-            }
-            if non_reparam_nodes:
-                downstream_costs, _ = _compute_downstream_costs(
-                    model_trace, guide_trace, non_reparam_nodes
+            # Find dependencies on non-reparameterizable sample sites for
+            # each cost term in the model and the guide.
+            model_deps, guide_deps = get_provenance(
+                eval_provenance(
+                    partial(
+                        track_nonreparam(get_importance_log_probs),
+                        seeded_model,
+                        seeded_guide,
+                        args,
+                        kwargs,
+                    )
                 )
+            )
 
             elbo = 0.0
+            # mapping from non-reparameterizable sample sites to cost terms influenced by each of them
+            downstream_costs = defaultdict(lambda: MultiFrameTensor())
             for site in model_trace.values():
                 if site["type"] == "sample":
                     elbo = elbo + jnp.sum(site["log_prob"])
+                    # add the log_prob to each non-reparam sample site upstream
+                    for key in model_deps[site["name"]]:
+                        downstream_costs[key].add(
+                            (site["cond_indep_stack"], site["log_prob"])
+                        )
             for name, site in guide_trace.items():
                 if site["type"] == "sample":
-                    log_prob_sum = jnp.sum(site["log_prob"])
-                    if name in non_reparam_nodes:
-                        surrogate = jnp.sum(
-                            site["log_prob"] * stop_gradient(downstream_costs[name])
-                        )
-                        log_prob_sum = (
-                            stop_gradient(log_prob_sum + surrogate) - surrogate
+                    elbo = elbo - jnp.sum(site["log_prob"])
```

Comment:
elbo = model_log_prob - guide_log_prob
current implementation: grad(elbo) = grad(model_log_prob) + grad(guide_surrogate)
this PR implementation: grad(elbo) = grad(model_log_prob) - grad(guide_log_prob) + grad(guide_surrogate)

```
if site["fn"].has_rsample:
     elbo = elbo - log_prob_sum
else:
     elbo = elbo - stop_gradient(log_prob_sum)
```